sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs

```diff
@@ -0,0 +1,347 @@
+﻿using Microsoft.IdentityModel.Tokens;
+using Microsoft.Rest;
+using System;
+using System.Collections.Generic;
+using System.IdentityModel.Tokens.Jwt;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.ContainerRegistry
+{
+    /*Robust handling of Basic and OAUTH2 authentication flows for the Azure Container Registry Runtime .Net SDK.
+     This class handles Basic Authentication as well as JWT token authentication using both username and password
+     routes as well as through exchanging AAD tokens. */
+    public class AcrClientCredentials : ServiceClientCredentials
+    {
+
+        #region Definitions
+        private class TokenCredentials : ServiceClientCredentials
+        {
+            private string AuthHeader { get; set; }
+
+            /*To be used for General Login Scheme*/
+            public TokenCredentials(string username, string password)
+            {
+                AuthHeader = EncodeTo64(username + ":" + password);
+            }
+            /*To be used for exchanging AAD Tokens for ACR Tokens*/
+            public TokenCredentials()
+            {
+                AuthHeader = null;
+            }
+            public override async Task ProcessHttpRequestAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+            {
+                if (request == null)
+                {
+                    throw new ArgumentNullException("request");
+                }
+                if (AuthHeader != null)
+                {
+                    request.Headers.Authorization = new AuthenticationHeaderValue("Basic", AuthHeader);
+                }
+                await base.ProcessHttpRequestAsync(request, cancellationToken);
+            }
+        }
+
+        public struct Token
+        {
+            public string TokenStr { get; set; }
+            public DateTime Expiration { get; set; }
+        }
+
+        private static readonly JwtSecurityTokenHandler JwtSecurityClient = new JwtSecurityTokenHandler();
+
+        public static Token MakeToken(string token)
+        {
+            Token tok = new Token();
+            tok.TokenStr = token;
+            SecurityToken fields = JwtSecurityClient.ReadToken(token);
+            tok.Expiration = fields.ValidTo;
+            return tok;
+        }
+
+        public enum LoginMode
```

Comment:
>LoginMode [](start = 20, length = 9)

Recommend moving to its own file outside the class

Reply:
I broke a lot of the token stuff apart into a separate class meant to handle tokens better.